### PR TITLE
Fixes Agent Metric

### PR DESCRIPTION
**Pull request by abachma2**
Attempts to fix the Agents metric, reported in Issue #174, by changing the exit variable read into the function and how the length of the table is determined.

### CODE REVIEW

File: cymetric/metrics.py [hunk 1]
```diff
@@ -274,19 +274,21 @@ def decommission_series(entry, exit):
...
     Furthermore, the Info table does not contain the AgentId column. This
     computation handles the calculation of the ExitTime in the face a
     significant amounts of missing data.
     """
     mergeon = ['SimId', 'AgentId']
-    ent = tools.raw_to_series(entry, ['SimId', 'AgentId'], 'Kind')
-    idx = ent.index
+    #ent = tools.raw_to_series(entry, ['SimId', 'AgentId'],'Kind')
+    #idx = ent.index
+    #idx = mergeon.size
```
Comment (bam241) [inline cymetric/metrics.py:287]: I think idx as to be defined as in l245:
`exit_index = ['SimId', 'ExitTime', 'Prototype']`

File: cymetric/metrics.py [hunk 2]
```diff
@@ -274,24 +274,24 @@ def decommission_series(entry, exit):
...
-    if exit is None:
-        agent_exit = pd.Series(index=idx, data=[np.nan] * len(idx))
+    if agent_exit is None:
+        agent_exit = entry[['SimId','AgentId']]
+        agent_exit['ExitTime'] = [np.nan]*len(agent_exit)
         agent_exit.name = 'ExitTime'
     else:
-        agent_exit = agent_exit.reindex(index=idx)
-    df = pd.merge(df, agent_exit.reset_index(), on=mergeon)
+        agent_exit = agent_exit#.reindex(index=idx)
```
Comment (bam241) [inline cymetric/metrics.py:293]: this else statement is not required anymore...

File: cymetric/metrics.py [hunk 3]
```diff
@@ -274,24 +274,24 @@ def decommission_series(entry, exit):
...
     mergeon = ['SimId', 'AgentId']
-    ent = tools.raw_to_series(entry, ['SimId', 'AgentId'], 'Kind')
-    idx = ent.index
+    idx = ['SimId','AgentId','ExitTime']
     df = entry[['SimId', 'AgentId', 'Kind', 'Spec', 'Prototype', 'ParentId',
                 'Lifetime', 'EnterTime']]
-    if exit is None:
-        agent_exit = pd.Series(index=idx, data=[np.nan] * len(idx))
+    if agent_exit is None:
+        agent_exit = entry[['SimId','AgentId']]
```
Comment (bam241) [inline cymetric/metrics.py:289]:
```suggestion
        agent_exit = entry[['SimId','AgentId']].copy()
```
Pandas seems to advices this usage instead.

File: cymetric/metrics.py [hunk 4]
```diff
@@ -285,12 +285,11 @@ def agents(entry, exit, decom, info):
...
     df = entry[['SimId', 'AgentId', 'Kind', 'Spec', 'Prototype', 'ParentId',
                 'Lifetime', 'EnterTime']]
-    df['ExitTime'] = [np.nan]*len(entry)
-    print(df)
+    df['Exits'] = [np.nan]*len(entry)
     if exit is not None:
         df = tools.merge_and_fillna_col(df, exit[['SimId','AgentId','ExitTime']], 
-                                        'ExitTime','ExitTime',on=mergeon)
-    print(df)
+                                        'Exits','ExitTime',how='left',on=mergeon)
```
Comment (bam241) [inline cymetric/metrics.py:291]:
```suggestion
                                        'Exits','ExitTime', on=mergeon)
```
this is probably not required.

File: cymetric/metrics.py [hunk 5]
```diff
@@ -282,16 +282,15 @@ def agents(entry, exit, decom, info):
...
-        agent_exit.name = 'ExitTime'
-    else:
-        agent_exit = agent_exit.reindex(index=idx)
-    df = pd.merge(df, agent_exit.reset_index(), on=mergeon)
+    df['ExitTime'] = [np.nan]*len(entry)
+    if exit is not None:
+        exit.columns = ['SimId','AgentId','Exits']
+        df = tools.merge_and_fillna_col(df, exit[['SimId','AgentId','Exits']], 
+                                        'ExitTime','Exits',on=mergeon)
+    #df.rename({'Exits':'ExitTime'},axis='columns')
```
Comment (bam241) [inline cymetric/metrics.py:293]: this is not used anymore.

File: cymetric/metrics.py [hunk 6]
```diff
@@ -282,16 +282,15 @@ def agents(entry, exit, decom, info):
     significant amounts of missing data.
     """
     mergeon = ['SimId', 'AgentId']
-    ent = tools.raw_to_series(entry, ['SimId', 'AgentId'], 'Kind')
-    idx = ent.index
+    idx = ['SimId','AgentId','ExitTime']
```
Comment (bam241) [inline cymetric/metrics.py:285]: I believe the idx isn't use...

**Review by bam241 — CHANGES_REQUESTED** [review]
Looking good!
Great job @abachma2 .
Some few clean-up and I'll be ready to merge !

File: cymetric/metrics.py [hunk 7]
```diff
@@ -282,16 +282,15 @@ def agents(entry, exit, decom, info):
...
-        agent_exit = pd.Series(index=idx, data=[np.nan] * len(idx))
-        agent_exit.name = 'ExitTime'
-    else:
-        agent_exit = agent_exit.reindex(index=idx)
-    df = pd.merge(df, agent_exit.reset_index(), on=mergeon)
+    df['ExitTime'] = [np.nan]*len(entry)
+    if exit is not None:
+        exit.columns = ['SimId','AgentId','Exits']
+        df = tools.merge_and_fillna_col(df, exit[['SimId','AgentId','Exits']], 
+                                        'ExitTime','Exits',on=mergeon)
```
Comment (bam241) [inline cymetric/metrics.py:291]: pep8 (our formatting guidelines) recommends space after commas, and there is trailing space at the end of the line.

Comment (bam241) [reply cymetric/metrics.py:291]: * pep8 is not our guideline but the one we try to follow :)

File: cymetric/metrics.py [hunk 8]
```diff
@@ -282,16 +282,13 @@ def agents(entry, exit, decom, info):
...
-    if exit is None:
-        agent_exit = pd.Series(index=idx, data=[np.nan] * len(idx))
-        agent_exit.name = 'ExitTime'
-    else:
-        agent_exit = agent_exit.reindex(index=idx)
-    df = pd.merge(df, agent_exit.reset_index(), on=mergeon)
+    df['ExitTime'] = [np.nan]*len(entry)
+    if exit is not None:
+        exit.columns = ['SimId', 'AgentId', 'Exits']
+        df = tools.merge_and_fillna_col(df, exit[['SimId','AgentId','Exits']],
```
Comment (bam241) [inline cymetric/metrics.py:290]:
```suggestion
        df = tools.merge_and_fillna_col(df, exit[['SimId', 'AgentId', 'Exits']],
```

File: cymetric/metrics.py [hunk 9]
```diff
@@ -282,16 +282,13 @@ def agents(entry, exit, decom, info):
...
+                                        'ExitTime','Exits',on=mergeon)
```
Comment (bam241) [inline cymetric/metrics.py:291]:
```suggestion
                                        'ExitTime', 'Exits', on=mergeon)
```

**Review by bam241 — COMMENTED** [review]
you missed 4 missing spaces. (I added them as suggestions)

**Review by bam241 — APPROVED** [review]
LGTM! 

Thank a lot @abachma2 for fixing this bug!

I'll merge this by the end of the week if no-one disagree

**Comment by abachma2** [comment]
Can this be merged now @bam241?

**Comment by bam241** [comment]
yes sorry for the delay!

**Comment by bam241** [comment]
Merging ! thx @abachma2 for fixing this !!!

**Pull request merged**